### PR TITLE
Docs for AutoBackbone & Backbone

### PR DESCRIPTION
**Pull request by merveenoyan**
This is the docs for `AutoBackbone` class and backbone. I'm also drafting a notebook but I will be off this week so it will be delayed.

### CODE REVIEW

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! Looking good 🤗 

Three general comments: 

1. We should use `out_indices` in at least one of the examples for layer selection. My preference would be for this to be used as the "primary" example. `out_features` requires the user to look at the model config to see the names of the features before selection, whereas `out_indices` you can just select the numbers directly. This also matches the behaviour in timm. 

2. There's a bit of magic (booo) that goes on between `out_features` and `out_indices`: 
 
* If one is passed in, then the other is set based on that e.g. if I passed `out_indices=(0,)` then `out_features` would be set to `"stage1"` and visa-versa. 
* If they're both passed in then they're verified to make sure they agree e.g. passing `out_indices=(0, 1, 2)` and `out_features=["stage4"]` would throw an error
* Passing in neither takes the last layer
* If I update one, then the other is updated e.g `config.out_indices = (-2, -1)` would then update `out_features` to `["stage3", "stage4"]` 

Not sure if this is something to highlight e.g. under a "sharp bits" kind of section?

3. We should highlight having access to all the timm checkpoints as backbones!

**Comment by merveenoyan** [comment]
@amyeroberts this didn't throw an error during inference:
```python
model = AutoBackbone.from_pretrained(
"microsoft/swin-tiny-patch4-window7-224", out_indices=(0,1,2), out_features=["stage4"]
)
```

I addressed rest of your comments. 🤗

**Comment by amyeroberts** [comment]
> @amyeroberts this didn't throw an error during inference:

Thanks for flagging - it should! I'll look into it 👀

File: docs/source/en/main_classes/backbones.md
Comment (MKhalusova) [inline docs/source/en/main_classes/backbones.md:26]: ".. get the feature maps for any stage." - any stage of what? It's a bit unclear.

Comment (MKhalusova) [reply docs/source/en/main_classes/backbones.md:26]: It would be cool to have an illustration here for your specific example. when the `out_indices=(0,)` what exactly does this mean?  It would make it much easier to understand the concept.

File: docs/source/en/main_classes/backbones.md
Comment (MKhalusova) [inline docs/source/en/main_classes/backbones.md:48]: I think it would be helpful to explain here what a "model neck configuration" is.

Comment (MKhalusova) [inline docs/source/en/main_classes/backbones.md:48]:
```suggestion
You can initialize architectures with model backbones by passing a model configuration to `backbone_config` to the model neck configuration. For example, below you can see how to initialize the [MaskFormer](../model_doc/maskformer) model with [ResNet](../model_doc/resnet) backbone.
```

File: docs/source/en/autoclass_tutorial.md
Comment (MKhalusova) [inline docs/source/en/autoclass_tutorial.md:148]:
```suggestion
`AutoBackbone` lets you use pretrained models as backbones and get feature maps as outputs from different stages of the models. Below you can see how to get feature maps from a [Swin](model_doc/swin) checkpoint.
```

File: docs/source/en/autoclass_tutorial.md
Comment (MKhalusova) [inline docs/source/en/autoclass_tutorial.md:146]: This guide is aimed at beginners, so it would be great to let them know when (or why) to use AutoBackbone.

**Review by MKhalusova — APPROVED** [review]
Thank you for working on these docs! I left a few comments and minor fixes.

**Review by amyeroberts — APPROVED** [review]
Awesome - thanks for iterating - looking great! :D 

+1 on @MKhalusova suggestion to have a diagram to demonstrate what a stage is.

**Comment by merveenoyan** [comment]
@MKhalusova I added illustrations and defined backbone/neck/head.

**Comment by amyeroberts** [comment]
> @amyeroberts this didn't throw an error during inference:
> 
> ```python
> model = AutoBackbone.from_pretrained(
> "microsoft/swin-tiny-patch4-window7-224", out_indices=(0,1,2), out_features=["stage4"]
> )
> ```
> 
> I addressed rest of your comments. 🤗

OK, so this is a bit tricky to resolve. The reason this is happening is: 
* The checkpoint `"microsoft/swin-tiny-patch4-window7-224"` was created before the swin backbone had been added. It therefore doesn't contain `out_features` or `out_indices` in the config. 
* When loading the model in `from_pretrained` the config is first created [from the json here](https://github.com/huggingface/transformers/blob/bd50402b56980ff17e957342ef69bd9b0dd45a7b/src/transformers/models/auto/configuration_auto.py#L1064) and then the kwargs (in this case `out_indices` and `out_features`) are [passed in here](https://github.com/huggingface/transformers/blob/bd50402b56980ff17e957342ef69bd9b0dd45a7b/src/transformers/models/auto/configuration_auto.py#L1081).
* In the config's `from_dict` method, the kwargs are applied [one at a time](https://github.com/huggingface/transformers/blob/bd50402b56980ff17e957342ef69bd9b0dd45a7b/src/transformers/configuration_utils.py#L766). This means that they're not verified against each other (but are still verified if they're consistent with `stage_names`).

This is part of the reason I'm wanting to deprecate `out_features` because handling this is annoying. I'll see if I can do something reasonable in the `BackboneConfigMixin` which isn't too convoluted to resolve.

File: docs/source/en/main_classes/backbones.md
Comment (MKhalusova) [inline docs/source/en/main_classes/backbones.md:69]:
```suggestion
In computer vision, models consist of backbone, neck, and a head. Backbone extracts the features, neck transforms the output of the backbone and head is used for the main task (e.g. object detection). You can initialize neck and head with model backbones by passing a model configuration to `backbone_config`. For example, below you can see how to initialize the [MaskFormer](../model_doc/maskformer) model with instance segmentation head with [ResNet](../model_doc/resnet) backbone.
```

**Review by MKhalusova — APPROVED** [review]
I think the visualization really helps to understand the stages, thank you for adding it! 
Once the CI is green, I think it's good to merge! 
Looks like one of the outputs in code examples doesn't match exactly what's expected.

**Comment by merveenoyan** [comment]
@amyeroberts since Maria has approved, can you merge?

**Comment by merveenoyan** [comment]
@MKhalusova Can you merge this if you can? Unfortunately, this wasn't included in recent release as I hope it would :/

**Comment by NielsRogge** [comment]
Hi,

Thanks a lot for this PR! 🙏 was wondering whether we could remove the out_features from the docs, since after offline discussion with @amyeroberts, we proposed to only favor `out_indices` from now on since it's easier to maintain. Could you open a follow-up PR regarding that?

**Pull request merged**
